src/utils/navigation.ts

```diff
@@ -44,8 +44,13 @@ export const isNavStateRestorable = (state: NavigationState) => {
   return latestRoute && !excludedRoutesFromRestoring.includes(latestRoute)
 }
 
-export const setNavigationState = (state: NavigationState) =>
-  rootStackNavigationRef.resetRoot(isNavStateRestorable(state) ? state : getInitialNavigationState())
+export const restoreNavigation = (navigation: NavigationProp<RootStackParamList>, state: NavigationState) => {
+  navigation.dispatch(CommonActions.reset(isNavStateRestorable(state) ? state : getInitialNavigationState()))
+}
 
-export const resetNavigationState = (initialRouteName?: string) =>
-  rootStackNavigationRef.resetRoot(getInitialNavigationState(initialRouteName))
+export const resetNavigation = (
+  navigation: NavigationProp<RootStackParamList>,
+  initialRouteName?: keyof RootStackParamList
+) => {
+  navigation.dispatch(CommonActions.reset(getInitialNavigationState(initialRouteName)))
+}
```

Comment:
Finally getting rid of the `rootStackNavigationRef.resetRoot` which were problematic and were causing unexpected issues. Our code is more type-safe now. We always require a valid navigation object before resetting/restoring the navigation.

Reply:
Yeah that feels good!